File: benchmark/ttt-template/tictactoe/board.py
"""tictactoe.board — core game logic (win / draw / move-validation).

This is the unit-tested surface of the generated app (see tests/test_board.py).
Pure stdlib, no third-party deps.
"""

from __future__ import annotations

from enum import Enum


class Player(str, Enum):
    X = "X"
    O = "O"

    @property
    def other(self) -> "Player":
        return Player.O if self is Player.X else Player.X


class GameState(str, Enum):
    IN_PROGRESS = "in_progress"
    X_WINS = "X_wins"
    O_WINS = "O_wins"
    DRAW = "draw"


# The 8 winning lines as index triples on the 0..8 flattened board.
_WIN_LINES: tuple[tuple[int, int, int], ...] = (
    (0, 1, 2), (3, 4, 5), (6, 7, 8),   # rows
    (0, 3, 6), (1, 4, 7), (2, 5, 8),   # cols
    (0, 4, 8), (2, 4, 6),              # diagonals
)


class InvalidMove(ValueError):
    """Raised when a move targets an out-of-range or occupied cell."""


class Board:
    """A 3x3 Tic-Tac-Toe board. Cells are None (empty) or a Player."""

    SIZE = 9

    def __init__(self) -> None:
        self._cells: list[Player | None] = [None] * self.SIZE
        self._turn: Player = Player.X

    # --- introspection -------------------------------------------------------
    @property
    def turn(self) -> Player:
        return self._turn

    def cell(self, index: int) -> Player | None:
        self._check_index(index)
        return self._cells[index]

    def empty_cells(self) -> list[int]:
        return [i for i, c in enumerate(self._cells) if c is None]

    def is_full(self) -> bool:
        return all(c is not None for c in self._cells)

    # --- move validation + application --------------------------------------
    @staticmethod
    def _check_index(index: int) -> None:
        if not isinstance(index, int) or not (0 <= index < Board.SIZE):
            raise InvalidMove(f"cell index out of range: {index!r} (expected 0..8)")

    def is_valid_move(self, index: int) -> bool:
        try:
            self._check_index(index)
        except InvalidMove:
            return False
        return self._cells[index] is None and self.state() is GameState.IN_PROGRESS

    def play(self, index: int) -> GameState:
        """Apply the current player's move at `index`; advance turn; return state.

        Raises InvalidMove for an out-of-range/occupied cell or a finished game.
        """
        self._check_index(index)
        if self.state() is not GameState.IN_PROGRESS:
            raise InvalidMove("game is already over")
        if self._cells[index] is not None:
            raise InvalidMove(f"cell {index} is already occupied")
        self._cells[index] = self._turn
        st = self.state()
        if st is GameState.IN_PROGRESS:
            self._turn = self._turn.other
        return st

    # --- win / draw detection ------------------------------------------------
    def winner(self) -> Player | None:
        for a, b, c in _WIN_LINES:
            v = self._cells[a]
            if v is not None and self._cells[b] is v and self._cells[c] is v:
                return v
        return None

    def state(self) -> GameState:
        w = self.winner()
        if w is Player.X:
            return GameState.X_WINS
        if w is Player.O:
            return GameState.O_WINS
        if self.is_full():
            return GameState.DRAW
        return GameState.IN_PROGRESS

    # --- rendering -----------------------------------------------------------
    def render(self) -> str:
        def sym(i: int) -> str:
            c = self._cells[i]
            return c.value if c is not None else str(i)
        rows = [
            f" {sym(0)} | {sym(1)} | {sym(2)} ",
            "---+---+---",
            f" {sym(3)} | {sym(4)} | {sym(5)} ",
            "---+---+---",
            f" {sym(6)} | {sym(7)} | {sym(8)} ",
        ]
        return "\n".join(rows)

File: benchmark/ttt-template/Sources/TicTacToeKit/Engine/Board.swift
/// A 3x3 Tic-Tac-Toe board. Cells are `nil` (empty) or a `Player`.
///
/// Ports `tictactoe/board.py` `Board` semantics exactly: move validation,
/// win/draw detection, and the fixed-width ASCII `render()` layout.
public struct Board: Sendable, Equatable {
    /// Number of cells on the board.
    public static let size = 9

    /// The 8 winning lines as index triples on the 0..8 flattened board.
    static let winLines: [(Int, Int, Int)] = [
        (0, 1, 2), (3, 4, 5), (6, 7, 8), // rows
        (0, 3, 6), (1, 4, 7), (2, 5, 8), // cols
        (0, 4, 8), (2, 4, 6), // diagonals
    ]

    private var cells: [Player?]
    private var currentTurn: Player

    public init() {
        cells = [Player?](repeating: nil, count: Board.size)
        currentTurn = .x
    }

    /// The player to move next.
    public var turn: Player { currentTurn }

    /// Validates `index` is an in-range cell index.
    private static func checkIndex(_ index: Int) throws(InvalidMove) {
        if index < 0 || index >= Board.size {
            throw InvalidMove.outOfRange(index)
        }
    }

    /// The contents of `index`, or `nil` if empty.
    /// - Throws: `InvalidMove.outOfRange` if `index` is not `0..<9`.
    public func cell(_ index: Int) throws(InvalidMove) -> Player? {
        try Board.checkIndex(index)
        return cells[index]
    }

    /// Indices of all empty cells, in ascending order.
    public func emptyCells() -> [Int] {
        cells.indices.filter { cells[$0] == nil }
    }

    /// Whether every cell is occupied.
    public func isFull() -> Bool {
        cells.allSatisfy { $0 != nil }
    }

    /// Whether playing at `index` right now would be legal (in-range, empty cell,
    /// game still in progress). Never throws — invalid indices simply return `false`.
    public func isValidMove(_ index: Int) -> Bool {
        guard index >= 0, index < Board.size else { return false }
        return cells[index] == nil && state() == .inProgress
    }

    /// Applies the current player's move at `index`, advances the turn if the
    /// game continues, and returns the resulting state.
    /// - Throws: `InvalidMove.outOfRange` for an out-of-range index,
    ///   `InvalidMove.gameOver` if the game already ended, or
    ///   `InvalidMove.occupied` if the cell is already taken.
    @discardableResult
    public mutating func play(_ index: Int) throws(InvalidMove) -> GameState {
        try Board.checkIndex(index)
        if state() != .inProgress {
            throw InvalidMove.gameOver
        }
        if cells[index] != nil {
            throw InvalidMove.occupied(index)
        }
        cells[index] = currentTurn
        let resulting = state()
        if resulting == .inProgress {
            currentTurn = currentTurn.other
        }
        return resulting
    }

    /// The winning player, or `nil` if there is no winner yet.
    public func winner() -> Player? {
        for (a, b, c) in Board.winLines {
            if let v = cells[a], cells[b] == v, cells[c] == v {
                return v
            }
        }
        return nil
    }

    /// The current terminal/non-terminal state of the board.
    public func state() -> GameState {
        switch winner() {
        case .x: return .xWins
        case .o: return .oWins
        case nil:
            return isFull() ? .draw : .inProgress
        }
    }

    /// Renders the board as fixed-width ASCII, matching the Python
    /// `Board.render()` output byte-for-byte.
    public func render() -> String {
        func sym(_ i: Int) -> String {
            cells[i]?.rawValue ?? String(i)
        }
        let rows = [
            " \(sym(0)) | \(sym(1)) | \(sym(2)) ",
            "---+---+---",
            " \(sym(3)) | \(sym(4)) | \(sym(5)) ",
            "---+---+---",
            " \(sym(6)) | \(sym(7)) | \(sym(8)) ",
        ]
        return rows.joined(separator: "\n")
    }
}
